Extracting render buffers and changing it to a more generic solution

BastiaanOlij opened the pull request:
Extracting and enhancing render buffers to be more generic and properly shield internals from the implementation so we can access and use this data from externals (in the long run).

We now have a base class `RenderSceneRender` in our main render implementation. This is now a storage class as it acts as a storage container for all buffers used by a viewport.
This class is subclassed in both the GLES3 renderer and RD renderer with implementations specific to those renderers. It is now also possible to implement this as a GDExtension class for a custom renderer.

The RD implementation adds functionality for named buffers so that effects and different RD implementations can create buffers with all logic already there for maintaining these buffers.

These buffers are named and organized in contexts. The contexts allows us to invalidate part of the buffers, say a setting for GI changes, instead of reallocating all buffers we will reallocate just the buffers within the GI context.

If properties on the viewport itself change, such as the size of the viewport or the scaling, all buffers are invalidated and destroyed and new buffers will be allocated as they are requested.

Finally, buffers can be marked as `unique`. This is a feature for later but will allow us to share buffers that are used as intermediate results and are defined the same over multiple viewports, to be shared to save memory. This is true by default as most buffers are only needed while we render a viewport and can thus be re-used when the next viewport is rendered. Only buffers that need to survive till the next frame (such as our `previous velocity` buffer in TAA) should have this setting false.

As such the API is designed for effects and render processes to always request the needed buffers each frame, if they do not exist they will be created after which they are cached and re-used in subsequent frames.

At the core of this implementation we find the following methods for our new texture buffers:

- `has_texture(p_context, p_buffer_name)` returns true if a texture exists
- `create_texture(p_context, p_buffer_name, p_data_format, p_usage_bits, p_texture_samples = TEXTURE_SAMPLES_1, p_size = Size2i(0, 0), uint32_t p_depth = 1, p_layers = 0, p_mipmaps = 1, p_view = RD::TextureView(), p_unique = true)` creates a texture and returns its RID. If a texture with this name already exists it is returned. Note that if no value is given for `p_size` or `p_layers` we use the (internal) size and view count given when `configure` was called.
- `create_texture_view(p_context, p_buffer_name, p_view_name, p_view)` creates a "view" on an existing texture (i.e. a view can do automatic sRGB conversion).
- `get_texture(p_context, p_buffer_name)` gets the RID of a texture previously created by `create_texture` or `create_texture_view`.
- `get_texture_format(p_context, p_buffer_name)` gets the format info which which a texture was previously created by `create_texture` or `create_texture_view` 
- `get_texture_slice(p_context, p_buffer_name, p_layer, p_mipmap)` retrieves a texture slice giving access to a specific layer/mipmap level of a texture. These slices are created and cached automatically. If a texture has only 1 layer and 1 mipmap (i.e. no mipmaps) this will return the main texture RID.
- `clear_context(p_context)` will invalidate (i.e. free) all textures for this context. 

And for data structures (objects subclassed from `RenderSceneBufferDataRD`):
- `has_data_buffer(p_name)` returns true if this data buffer exists
- `set_data_buffer(p_name, p_buffer)` sets our data buffer
- `get_data_buffer(p_name)` gets our data buffer

Note: while the idea in the long run will be that we can also build an Render Device based renderer in GDExtension, there is still some work to do before `RenderSceneRenderRD` is fully callable from GDExtension. 

This PR also changes many effects to use this new system and splits out the remaining effects that haven't undergone this treatment.

So far done:
- TAA
- VRS
- FSR (also fixing #64254)
- Sub surface scattering
- GI

Still to split out:
- Copy
- Luminance
- Roughness limiter

Still need to convert to using the new buffer approach:
- SSAO
- SSIL
- SSR
- DOF
- FOG
- SKY
-

BastiaanOlij committed:
Extracting render buffers and changing it to a more generic solution

BastiaanOlij commented:
@clayjohn There might be a miss in the configure code in GLES3, I had a weird rebase issue where it seemed to have thought some code changed but since its moved to a new class auto-merge doesn't like it.

On that subject, I'm thinking of implementing `RenderBuffers::configure` on the base class so it purely captures the parameters, as both GLES3, RD and any other implementation will likely store these anyway. Together with access methods and adding `_bind_methods` we can make it all accessible.

That said, there is a good argument for keeping `RenderBuffers` a purely virtual class and then implementing a `RenderBuffersExtension` for GDExtension

JFonS commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -1736,76 +1567,85 @@ void RendererSceneRenderRD::_process_ssil(RID p_render_buffers, RID p_environmen
-		tf.height = rb->internal_height;
-		tf.array_layers = rb->view_count; // create a layer for every view
-		tf.usage_bits = RD::TEXTURE_USAGE_SAMPLING_BIT | (_render_buffers_can_be_storage() ? RD::TEXTURE_USAGE_STORAGE_BIT : 0);
+	if (!p_render_buffers->has_texture(SNAME("taa"), SNAME("taa_history"))) {
+		uint32_t usage_bits = RD::TEXTURE_USAGE_SAMPLING_BIT | (_render_buffers_can_be_storage() ? RD::TEXTURE_USAGE_STORAGE_BIT : 0);
+
+		RD::DataFormat format = _render_buffers_get_color_format();
+		p_render_buffers->create_texture(SNAME("taa"), SNAME("taa_history"), format, usage_bits);
+		p_render_buffers->create_texture(SNAME("taa"), SNAME("taa_temp"), format, usage_bits);
```
Stylistic nitpick, but we could remove `taa_` bit of the texture names, since it's already accessed by the `"taa"` prefix.

BastiaanOlij replied:
Will do!

JFonS reviewed (approved):
I really like the clean-up. It makes working with render buffers much nicer.

I went over the TAA bits and everything seems fine, I just have a tiny nitpick.

BastiaanOlij commented:
Ok, in reworking FSR I've also fixed a bunch of things here and figured this never properly worked, mixing up which buffers to read from and which to write. I think there is more room for improving the code around setting up FSR but for now I'm happy with how things are.

It results in all 3D rendering happening at the lower resolution including SSAO, SSIL, SSR, Glow, DOF, tonemapping, etc.

FSR is applied at the very end upscaling the result to the correct resolution and writing directly into our render target. 

The only issue here is that FSR uses a compute shader which means our render target requires `TEXTURE_USAGE_STORAGE_BIT`, while only FSR uses it as such on GPUs that do not support compute shaders setting this bit may be an issue.
I'm not sure of the best way to solve this, probably the same way as we did for VRS. 
So in `TextureStorage::_update_render_target` this line of code needs to become optional: `rd_format.usage_bits |= RD::TEXTURE_USAGE_STORAGE_BIT; // FIXME we need this only when FSR is enabled`

As a bonus, FSR should work in VR now as well :P (at least with the desktop renderer)

@clayjohn , @Calinou would really appreciate if you guys can check what I've done and see if you agree, and test.

BastiaanOlij commented:
Subsurface scattering has been done now as well and should support stereoscopic. Not sure if there needs to be any adjustment for the camera. I think it can use just the standard one.

clayjohn reviewed (commented):
The general approach looks fine to me. 

I can think of two things that will likely require further discussion:
1. What resources can be moved out of RenderBuffers (i.e. can we move SSEffects out? If not, can we use the more dynamic resource creation like we do for the RenderBuffer owned resources (i.e. internal texture)?
2. How much of the reorganization should be done in this PR vs another. I think you have quite a bit more work ahead of you to get all the effects and things moved over to the new API. It may make sense to merge this and then continue the work to avoid conflicting too much with other work going on

BastiaanOlij commented:
> The general approach looks fine to me.
> 
> I can think of two things that will likely require further discussion:
> 
>     1. What resources can be moved out of RenderBuffers (i.e. can we move SSEffects out? If not, can we use the more dynamic resource creation like we do for the RenderBuffer owned resources (i.e. internal texture)?
> 
>     2. How much of the reorganization should be done in this PR vs another. I think you have quite a bit more work ahead of you to get all the effects and things moved over to the new API. It may make sense to merge this and then continue the work to avoid conflicting too much with other work going on

For 1, I want to solve all of it because I do not like the hardcoded nature of what is added to RenderBuffers, it makes it impossible to add new effects or new implementation of effects needed for other render device based implementations. This may mean adding a structure to render buffers that allows registering data structures.

For 2, this is something I've been thinking a lot about lately, now that the foundation is there it makes sense to move to merging this sooner rather than later and then doing the finishing touches in separate PRs per part.

BastiaanOlij commented:
Ok, just cleaned up a lot of the code in the mobile and clustered renderer around buffer creation and switched things over to use our cached framebuffer. This should prevent buffers from being created that are unused and opens the door for our mobile renderer to use our swap chain as the final destination.

Both render implementations now subclass their data object from the new `RenderSceneBufferDataRD`. This new class allows us to register named data objects with our render buffer class. This has been designed to also work for other data buffers such as those used by some of our effects.

BastiaanOlij commented:
Hmm, seeing I'm using unsigned ints the CI seems to dislike `ERR_FAIL_INDEX_V` checking for < 0...

akien-mga commented:
Use `ERR_FAILED_UNSIGNED_INDEX_V`.

BastiaanOlij commented:
> `ERR_FAIL_UNSIGNED_INDEX_V`

How did I not see we have that :P

clayjohn commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp:
```diff
@@ -0,0 +1,486 @@
+
+void RenderSceneBuffersRD::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("has_texture", "context", "name"), &RenderSceneBuffersRD::has_texture);
+	// FIXME we can't pass RD::DataFormat, RD::TextureSamples and RD::TextureView in ClassDB, need to solve views differently...
```
We may just need to expose these to scripting. We have exposed many RD-specific types already, like Uniforms, etc.

BastiaanOlij replied:
Yeah I think it's just the TextureView property that is a real issue because this is a structure, not a variable...

clayjohn commented on servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:
```diff
@@ -69,183 +69,175 @@ void RenderForwardMobile::_map_forward_id(ForwardIDType p_type, ForwardID p_id,
-	if (color_msaa.is_valid()) {
-		RD::get_singleton()->free(color_msaa);
-		color_msaa = RID();
+void RenderForwardMobile::RenderBufferDataForwardMobile::free_data() {
```
This appears to be leading to the crash caught by CI. Not sure what is going on, but it should be fixed before merging

BastiaanOlij replied:
I'll have a look, I did some last minute changes here so I must have missed it being called before configure was called the first time (or indeed, being called from configure for the first time).

BastiaanOlij replied:
Ah forgot to default a pointer to `nullptr`, so lovely that MSVC does this automatically so it's not habitual yet...

reduz commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.h:
```diff
@@ -0,0 +1,273 @@
+	void clear_context(const StringName &p_context);
+
+	// Data buffers
+	bool has_data_buffer(const StringName &p_name);
```
I think I would rename this to custom_data / RenderBuffersCustomData, because in general this can contain more things than just buffers.

BastiaanOlij replied:
More than happy to apply the new names:)

reduz commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.h:
```diff
@@ -0,0 +1,273 @@
+
+	// info from our renderer
+	void set_can_be_storage(const bool p_can_be_storage) { can_be_storage = p_can_be_storage; }
+	void set_max_cluster_elements(const uint32_t p_max_elements) { max_cluster_elements = p_max_elements; }
```
Cluster should be move to a CustomData I guess too.

BastiaanOlij replied:
Yup, but I think there is a fair amount of work left to extract the common logic from the cluster builder that shouldn't have been in the cluster builder to begin with, so that one I want to leave for a separate PR.

reduz commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.h:
```diff
@@ -0,0 +1,273 @@
+		return get_texture_slice(SNAME("RenderSceneBuffersRD"), SNAME("depth_texture"), p_layer, 0);
+	}
+
+	// back buffer (color + depth)
```
I would probably move these and the ones below away from here since they will only be created when they make sense. If they are not around when needed, and something requests it and its not found, then it should be considered an error and its fine.

BastiaanOlij replied:
You mean the back buffer? yeah I can improve on that :)

BastiaanOlij replied:
The depth back buffer now creates the buffers if needed in the copy function and where the back buffer is used it's accessed directly.

For the color buffer I still need to redo the blur approach.

reduz commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.h:
```diff
@@ -0,0 +1,273 @@
+	// Velocity, currently only used by TAA (Clustered) but we'll be using this in other places soon too.
+
+	void ensure_velocity();
+	bool has_velocity_buffer(bool p_has_msaa) { return has_texture(SNAME("RenderSceneBuffersRD"), p_has_msaa ? SNAME("msaa_velocity") : SNAME("velocity")); }
```
I think for the categories and commonly used textures, it should be fine to have a .h that does defines like:

```c++
#define RB_SCOPE_BUFFERS SNAME("render_buffers")
#define RB_SCOPE_VRS SNAME("VRS")
// and
#define RB_TEX_VELOCITY SNAME("velocity")
#define RB_TEX_COLOR SNAME("color")
#define RB_TEX_DEPTH SNAME("depth")

// so you can do like:
rb->get_texture(RB_SCOPE_BUFFERS,RB_TEX_COLOR);
rb->get_texture(RB_SCOPE_BUFFERS,RB_TEX_DEPTH);
```
Just for the sake of not using strings if not needed. They can be defined locally also in cpp files for custom effects and the likes, so no need to make all global.

BastiaanOlij replied:
Ah we discussed this on chat and I actually moved away from this after our discussion. I agree, having defines makes it much less likely to make mistakes.

reduz commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.h:
```diff
@@ -0,0 +1,273 @@
+	_FORCE_INLINE_ RID get_vrs_texture() const { return get_texture(SNAME("VRS"), SNAME("Texture")); }
+	_FORCE_INLINE_ RID get_vrs_texture(const uint32_t p_layer) { return get_texture_slice(SNAME("VRS"), SNAME("Texture"), p_layer, 0); }
+
+	////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
I guess this will all move to generic API right?

BastiaanOlij replied:
The problem with VRS is that it needs to be used when setting up frame buffers and is thus handled by the main class. I want to give this some more thought how we can better have the implementation just in the VRS effects class.

BastiaanOlij replied:
Ok I changed how this works so we not longer have these getters. Where VRS is used it just accesses the buffers directly.

reduz commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.h:
```diff
@@ -0,0 +1,273 @@
+	// Named Textures
+
+	bool has_texture(const StringName &p_context, const StringName &p_texture_name) const;
+	RID create_texture(const StringName &p_context, const StringName &p_texture_name, const RD::DataFormat p_data_format, const uint32_t p_usage_bits, const RD::TextureSamples p_texture_samples = RD::TEXTURE_SAMPLES_1, const Size2i p_size = Size2i(0, 0), const uint32_t p_depth = 1, const uint32_t p_layers = 0, const uint32_t p_mipmaps = 1, RD::TextureView p_view = RD::TextureView(), bool p_unique =  ...
```
Feel free to pass `RD::TextureFormat`, we have RDTextureFormat object that can be used for the binder specific version, this can be added later on when exposed.

BastiaanOlij replied:
The problem we have at the moment is that in many cases we have the RD::DataFormat, not the RD::TextureFormat. So we would need to do a little more work to make sure we use TextureFormat more consistantly.

reduz replied:
I have no idea when we use this, but in this case you definitely want to pass TextureFormat, as this may change in the future versions of Godot and force us to modify this API and break compatibility.

BastiaanOlij replied:
This all became an issue when we introduced the mobile renderer and many of the buffers needed to be 32bit buffers instead of 64bit ones. And we're already using RD::DataFormat there.

But in hindsight, we probably should have used RD::TextureFormat there as well. I'll see how easy it is to change over fully.

BastiaanOlij replied:
Ah hang on, the penny drops. You want to pass the `RD::TextureFormat` struct in its entirety? For some reason I had in my head that you wanted to change this to `Image::Format` or something similar that was already exposed to GDScript...

This would greatly bloat a lot of the code where we now have a single line solution but I can see the use case for having it. 

For now I've changed it so that we have a:
`RID create_texture_from_format(const StringName &p_context, const StringName &p_texture_name, const RD::TextureFormat &p_texture_format, RD::TextureView p_view = RD::TextureView(), bool p_unique = true);`
That takes a texture format struct as input.

But I've left:
`RID create_texture(const StringName &p_context, const StringName &p_texture_name, const RD::DataFormat p_data_format, const uint32_t p_usage_bits, const RD::TextureSamples p_texture_samples = RD::TEXTURE_SAMPLES_1, const Size2i p_size = Size2i(0, 0), const uint32_t p_layers = 0, const uint32_t p_mipmaps = 1, bool p_unique = true);`
in place and this is what is called in most places with most properties left to default.
Internally it builds the correct texture format and calls `create_texture_from_format`.

This gives us a best of both world solution, it keeps the API lean in 99% of the cases, and in the 1% where we need to define a specific texture format, we can do so.

reduz commented on servers/rendering/renderer_scene_cull.h:
```diff
@@ -1054,10 +1054,10 @@ class RendererSceneCull : public RendererScene {
+	void render_empty_scene(Ref<RenderSceneBuffers> p_render_buffers, RID p_scenario, RID p_shadow_atlas);
 
-	void render_camera(RID p_render_buffers, RID p_camera, RID p_scenario, RID p_viewport, Size2 p_viewport_size, bool p_use_taa, float p_screen_mesh_lod_threshold, RID p_shadow_atlas, Ref<XRInterface> &p_xr_interface, RendererScene::RenderInfo *r_render_info = nullptr);
+	void render_camera(Ref<RenderSceneBuffers> p_render_buffers, RID p_camera, RID p_scenario, RID p_viewport, Size2 p_viewport_size, bool p_use_taa, float p_screen_mesh_lod_threshold, RID p_shadow_atlas, Ref<XRInterface> &p_xr_interface, RendererScene::RenderInfo *r_render_info = nullptr);
```
Super minor nitpick but we should likely pass const Ref<>& here and in similar places.

BastiaanOlij replied:
Owh totally makes sense

reduz commented on servers/rendering/storage/render_scene_buffers.h:
```diff
@@ -0,0 +1,60 @@
+	RenderSceneBuffers(){};
+	virtual ~RenderSceneBuffers(){};
+
+	virtual void configure(RID p_render_target, const Size2i p_internal_size, const Size2i p_target_size, float p_fsr_sharpness, float p_texture_mipmap_bias, RS::ViewportMSAA p_msaa, RenderingServer::ViewportScreenSpaceAA p_screen_space_aa, bool p_use_taa, bool p_use_debanding, uint32_t p_view_count);
```
I think the ones you pass below should not be repeated above.

BastiaanOlij replied:
I'm not sure I follow you ? :)

BastiaanOlij replied:
Ah you mean like `set_fsr_sharpness`. I added those in so the buffers don't get rebuild when the sharpness changes. Before configure would always get called even if we change settings that are just inputs to shaders... I wonder if those even should be in RenderBuffers.

reduz commented:
Looking amazing, Bastiaan!

reduz commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -1672,23 +1672,23 @@ void RasterizerSceneGLES3::_setup_lights(const RenderDataGLES3 *p_render_data, b
 }
 
-void RasterizerSceneGLES3::render_scene(RID p_render_buffers, const CameraData *p_camera_data, const CameraData *p_prev_camera_data, const PagedArray<RenderGeometryInstance *> &p_instances, const PagedArray<RID> &p_lights, const PagedArray<RID> &p_reflection_probes, const PagedArray<RID> &p_voxel_gi_instances, const PagedArray<RID> &p_decals, const PagedArray<RID> &p_lightmaps, const PagedArray<R ...
+void RasterizerSceneGLES3::render_scene(Ref<RenderSceneBuffers> &p_render_buffers, const CameraData *p_camera_data, const CameraData *p_prev_camera_data, const PagedArray<RenderGeometryInstance *> &p_instances, const PagedArray<RID> &p_lights, const PagedArray<RID> &p_reflection_probes, const PagedArray<RID> &p_voxel_gi_instances, const PagedArray<RID> &p_decals, const PagedArray<RID> &p_lightmap ...
```
missing the const part, should be const Ref<RenderSceneBuffers> &
you can then assign to something non const again and it works fine, its just for the compiler to do the right optimization.


That said, given the way you use it, maybe its indeed better to just pass it by value.

BastiaanOlij replied:
Ah oops, changed them all (I think)

reduz commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h:
```diff
@@ -33,12 +33,22 @@
 #include "servers/rendering/renderer_rd/shaders/scene_forward_clustered.glsl.gen.h"
 #include "servers/rendering/renderer_rd/storage_rd/utilities.h"
 
+#define RB_SCOPE_CLUSTERED SNAME("clustered")
```
Keep in mind we will have forward and deferred clustered eventually and they don't use the same buffers.

BastiaanOlij replied:
Made this more explicit: `#define RB_SCOPE_FORWARD_CLUSTERED SNAME("forward_clustered")`

BastiaanOlij commented:
@Geometror I had some interesting merge conflicts with your recently merged 2D MSAA PR, could you check if things still work correctly for you on this PR or if I inadvertently undid one of your changes?

BastiaanOlij commented:
Ok, pretty much complete changing over the logic with blur textures however found out there some shader issues snuck in so still debugging those...

BastiaanOlij commented:
Ok, found the bug with VRS and found the bug with the mobile renderer. All seems to be working fine now. Changed the blur texture code over to using our new approach.

Geometror commented:
@BastiaanOlij Tested 2D/3D MSAA and it looks like it still works, even if configured separately. However, while testing I found out that enabling 3D MSAA together with TAA fails with this PR.
```
ERROR: Condition "!E" is true. Returning: TEXTURE_SAMPLES_1
   at: RenderingDeviceVulkan::framebuffer_format_get_texture_samples (drivers\vulkan\rendering_device_vulkan.cpp:4081)
ERROR: Mismatch fragment shader output mask (5) and framebuffer color output mask (0) when binding both in render pipeline.
   at: (drivers\vulkan\rendering_device_vulkan.cpp:6463)
ERROR: Condition "pipeline.is_null()" is true. Returning: RID()
   at: PipelineCacheRD::_generate_version (servers\rendering\renderer_rd\pipeline_cache_rd.cpp:60)
ERROR: Condition "!dl" is true.
   at: RenderingDeviceVulkan::draw_list_bind_uniform_set (drivers\vulkan\rendering_device_vulkan.cpp:7640)
ERROR: Condition "!dl" is true.
   at: RenderingDeviceVulkan::draw_list_set_push_constant (drivers\vulkan\rendering_device_vulkan.cpp:7747)
ERROR: Condition "!dl" is true.
   at: RenderingDeviceVulkan::draw_list_draw (drivers\vulkan\rendering_device_vulkan.cpp:7765)
ERROR: Immediate draw list is already inactive.
   at: (drivers\vulkan\rendering_device_vulkan.cpp:8068)
ERROR: Source texture can't be copied while a render pass that uses it is being created. Ensure render pass is finalized (and that it was created with RENDER_PASS_CONTENTS_FINISH) to unbind this texture.
   at: (drivers\vulkan\rendering_device_vulkan.cpp:3055)
ERROR: Destination texture requires the TEXTURE_USAGE_CAN_COPY_TO_BIT in order to be retrieved.
   at: (drivers\vulkan\rendering_device_vulkan.cpp:3068)
```

BastiaanOlij commented:
@Geometror cheers! I'll have a look, looks like it's not setting up the framebuffer correctly

reduz commented:
I think it should be good for now, I think we can further clean up later

akien-mga commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp:
```diff
@@ -0,0 +1,559 @@
+		uint32_t usage_bits = RD::TEXTURE_USAGE_COLOR_ATTACHMENT_BIT | RD::TEXTURE_USAGE_SAMPLING_BIT | RD::TEXTURE_USAGE_STORAGE_BIT;
+
+		if (msaa_3d != RS::VIEWPORT_MSAA_DISABLED) {
+			uint32_t msaa_usage_bits = usage_bits || RD::TEXTURE_USAGE_CAN_COPY_FROM_BIT;
```
Wrong operator :)

BastiaanOlij replied:
yup, that was the one :)

BastiaanOlij commented:
@Geometror ok thing you found should be fixed :)

Geometror commented:
@BastiaanOlij I can confirm that everything seems to work now. Great work!

akien-mga commented:
Thanks!
